### PR TITLE
Add ibv_query_qp_data_in_order flags

### PR DESCRIPTION
**Pull request by mrgolin**
Add definition for ibv_query_qp_data_in_order flags to allow querying for specific capabilities related to data polling support, currently adding 128 bytes in order flag (LL128).
Adopt existing code and add support in EFA provider.

### CODE REVIEW

**Comment by mrgolin** [comment]
Related kernel patch:
https://lore.kernel.org/linux-rdma/20230219081328.10419-1-mrgolin@amazon.com/T/#u

**Comment by mrgolin** [comment]
@rleon I would really appreciate if this can be reviewed for the upcoming release.

**Comment by rleon** [comment]
> @rleon I would really appreciate if this can be reviewed for the upcoming release.

You posted kernel patch day before merge window starts. I afraid that it is too late for UAPI changes.

**Comment by mrgolin** [comment]
> > @rleon I would really appreciate if this can be reviewed for the upcoming release.
> 
> You posted kernel patch day before merge window starts. I afraid that it is too late for UAPI changes.

Thanks Leon I understand the timeline and I wish we could submit this earlier. Yet, the kernel patch is really simple one and it changes UAPI only by a definition of another bit in an existing field.

**Comment by rleon** [comment]
> > > @rleon I would really appreciate if this can be reviewed for the upcoming release.
> > 
> > 
> > You posted kernel patch day before merge window starts. I afraid that it is too late for UAPI changes.
> 
> Thanks Leon I understand the timeline and I wish we could submit this earlier. Yet, the kernel patch is really simple one and it changes UAPI only by a definition of another bit in an existing field.

I'm aware. however I don't want to take any chances and get angry about sending PR to Linus with patches not being in linux-next.

Thanks

**Comment by rleon** [comment]
@jgunthorpe , you are sending PRs to Linus, so it is your call.

Thanks

**Comment by mrgolin** [comment]
> This doesn't make alot of sense to me.. "in order" reflects the entire message stream, so what does "within a block" even mean? It seems to be some completely orthogonal concept related to cache tearing not message ordering?

From documentation:
```
**ibv_query_qp_data_in_order()** Checks whether WQE data is guaranteed to be
written in-order, and thus reader may poll for data instead of poll for completion.
This function indicates data is written in-order within each WQE, but cannot be used to determine ordering between separate WQEs.
```
This function is used to determine whether data can be polled, i.e. it's guaranteed that if some byte was written then all previous bytes of this message are ready to be used.
Some devices (e.g. EFA) may not support this for the full message but can guarantee write in order inside each data block of a particular size.

**Comment by mrgolin** [comment]
> > This doesn't make alot of sense to me.. "in order" reflects the entire message stream, so what does "within a block" even mean? It seems to be some completely orthogonal concept related to cache tearing not message ordering?
> 
> From documentation:
> 
> ```
> **ibv_query_qp_data_in_order()** Checks whether WQE data is guaranteed to be
> written in-order, and thus reader may poll for data instead of poll for completion.
> This function indicates data is written in-order within each WQE, but cannot be used to determine ordering between separate WQEs.
> ```
> 
> This function is used to determine whether data can be polled, i.e. it's guaranteed that if some byte was written then all previous bytes of this message are ready to be used. Some devices (e.g. EFA) may not support this for the full message but can guarantee write in order inside each data block of a particular size.

@jgunthorpe Does it make sense now?

**Comment by YonatanNachum** [comment]
We have a kernel patch: [link](https://patchwork.kernel.org/project/linux-rdma/patch/20230219081328.10419-1-mrgolin@amazon.com/) 
Thanks.

File: libibverbs/driver.h
Comment (gal-pressman) [inline libibverbs/driver.h:424]: I would keep the flags here, the provider can choose to ignore them.

File: libibverbs/verbs.c
Comment (gal-pressman) [inline libibverbs/verbs.c:713]: You can use check_comp_mask() here

File: libibverbs/verbs.c
Comment (gal-pressman) [inline libibverbs/verbs.c:705]: Not sure comp_mask is the best name for this variable.

**Comment by gal-pressman** [comment]
IIUC, it means every 128 bytes are written "atomically", it doesn't really replace the completion though, as you have no idea when the entire message is completed (unless you check each 128 chunk separately?).

What is this used for?

File: libibverbs/driver.h
Comment (mrgolin) [reply libibverbs/driver.h:424]: The idea of this interface change is that providers will now return all their supported capabilities (related to data inorder) without depending on the request. I think flags here may be confusing, when would you use it in provider code?

**Comment by gal-pressman** [comment]
Makes sense.
LL128 is where you have a flags header for each 120 bytes of data, right?

**Comment by mrgolin** [comment]
> Makes sense. LL128 is where you have a flags header for each 120 bytes of data, right?

Right, 120 bytes of data and 8 bytes flag.

File: libibverbs/driver.h
Comment (gal-pressman) [reply libibverbs/driver.h:424]: I don't know, depends on the flags.
Anyway, it's not a big deal, I guess the flags can be reintroduced if/when they're needed.

File: libibverbs/verbs.c
Comment (mrgolin) [reply libibverbs/verbs.c]: Done, thanks.

**Review by jgunthorpe — CHANGES_REQUESTED** [review]

File: libibverbs/verbs.h
Comment (mrgolin) [reply libibverbs/verbs.h:3214]: Do you suggest that passing IBV_QUERY_QP_DATA_RETURN_FLAGS as flags to that function will make it return supported capabilities instead of 0/1?

Comment (mrgolin) [reply libibverbs/verbs.h:3214]: That actually was one of the options we considered but it didn't feel as a natural API extension but more like combining two into a single one. Is there any future use case for this function you can think of that won't be satisfied by querying for a capability bit?

Comment (mrgolin) [reply libibverbs/verbs.h]: @jgunthorpe Changed according to your request.

File: kernel-headers/rdma/efa-abi.h
Comment (wenduwan) [inline kernel-headers/rdma/efa-abi.h:3]: Same as below, the latest open source policy advises dateless notices, e.g.
```
Copyright Amazon.com, Inc. or its affiliates. All rights reserved.
```

File: libibverbs/verbs.h
Comment (mrgolin) [reply libibverbs/verbs.h]: If it looks ok now, can we get this merged?

File: libibverbs/verbs.c
Comment (jgunthorpe) [inline libibverbs/verbs.c:712]: There is only one other implementation just fix it to return the new style always

**Review by jgunthorpe — CHANGES_REQUESTED** [review]
The rest looks OK

File: libibverbs/verbs.c
Comment (mrgolin) [reply libibverbs/verbs.c]: Ok, done.

File: providers/efa/verbs.c
Comment (mrgolin) [reply providers/efa/verbs.c:1533]: So to make sure we are aligned, you would remove flags param from provider ops as I did in the first place or just move this condition to common code and ignore flags in providers?

Comment (jgunthorpe) [reply providers/efa/verbs.c:1533]: yes just ignore the flag

Comment (mrgolin) [reply providers/efa/verbs.c]: Done

File: kernel-headers/rdma/efa-abi.h
Comment (mrgolin) [reply kernel-headers/rdma/efa-abi.h:3]: @wenduwan Thanks.
This is a kernel header, can't change it here so will leave dated copyrights for now.

**Comment by mrgolin** [comment]
@jgunthorpe any additional comments?

**Comment by jgunthorpe** [comment]
I'm still a bit nervous about this - AFAICT the only way to achieve this in a modern system is if the device promises to generate 128 byte PCIe MemWr TLPs so that the platform can execute single TLPs "in order". But the max TLP size is controlled by PCI config space, so how can verbs know it is 128 bytes at this point?

The existing flag is basically saying 'the device does not use relaxed ordering and writes all bytes in any TLPs in order" and further that "the platform does not re-order non-relaxed ordering TLPs", which doesn't have this problem.

And further, semantically, according to verbs MRs created should be of the non-relaxed ordering type anyhow, so how do you get into a situation where otherwise in-order TLPs are re-ordered? Does this only work with relaxed ordering or is something in EFA wrongly forcing relaxed ordering in the MRs?

**Comment by mrgolin** [comment]
> I'm still a bit nervous about this - AFAICT the only way to achieve this in a modern system is if the device promises to generate 128 byte PCIe MemWr TLPs so that the platform can execute single TLPs "in order". But the max TLP size is controlled by PCI config space, so how can verbs know it is 128 bytes at this point?
> 
> The existing flag is basically saying 'the device does not use relaxed ordering and writes all bytes in any TLPs in order" and further that "the platform does not re-order non-relaxed ordering TLPs", which doesn't have this problem.
 
Whenever 128 bytes in-order is promised it's up to the device to make sure each such block isn't getting splitted and reordered nether on the device/communication level nor by PCIe TLPs. It can achieve the PCIe part as you suggested by not using PCIe relaxed ordering (same way it is done for whole message in-order support) or if it is familiar with the platform, ensure that write TLPs don't split data anywhere in-between 128 bytes boundaries.
Either way it is less strict than whole message in-order so I believe that other providers may also correctly support this.

> And further, semantically, according to verbs MRs created should be of the non-relaxed ordering type anyhow, so how do you get into a situation where otherwise in-order TLPs are re-ordered? Does this only work with relaxed ordering or is something in EFA wrongly forcing relaxed ordering in the MRs?

I think it is mostly valuable for relaxed ordering to gain both from relaxed order performance improvement and from LL128 data polling low latency but if 128 bytes in-order is assured for relaxed ordering I don't see a reason why it won't be for non-relaxed ordering.

**Comment by jgunthorpe** [comment]
Having just learned about this NCCL LL128 stuff, this patch is somewhat wrong :(

The query_qp_data_in_order() is all about the *CPU*'s perception of ordering. This new flag you've added is really about the device's ability to generate 128 byte PCIe TLPs.

**Comment by mrgolin** [comment]
I believe this verb documentation can, hopefully not too late for some users, be improved to more explicitly state where its responsibility ends but looking at the discussion on original [PR](https://github.com/linux-rdma/rdma-core/pull/1009) that introduced this there wasn't any intension to cover destination memory behavior. For example your comment from that discussion:
>Sean, I covered that above, this API is only about the behavior up until verbs delivers PCIe TLPs out of the HCA, it doesn't try very hard to cover platform behavior. MR specific configuration including relaxed ordering and the behavior of the target memory is out of scope because we can't get that kind of information from verbs.

So I think what we did here perfectly aligns with this, extending it to support additional cases.

**Pull request merged**
